nixos/modules/services/web-apps/nextcloud-notify_push.nix

```diff
@@ -0,0 +1,57 @@
+{ config, lib, pkgs, ... }:
+
+let
+  cfg = config.services.nextcloud.notify_push;
+in
+{
+  options.services.nextcloud.notify_push = {
+    enable = lib.mkEnableOption (lib.mkDoc "Notify push");
+    package = lib.mkOption {
+      type = lib.types.package;
+      default = pkgs.nextcloud-notify_push;
+      defaultText = lib.literalMD "pkgs.nextcloud-notify_push";
+      description = lib.mdDoc "Which package to use for notify_push";
+    };
+    socketPath = lib.mkOption {
+      type = lib.types.str;
+      default = "/run/nextcloud-notify_push/sock";
+      description = lib.mdDoc "Socket path to use for notify_push";
+    };
+    logLevel = lib.mkOption {
+      type = lib.types.enum [ "error" "warn" "info" "debug" "trace" ];
+      default = "error";
+      description = lib.mdDoc "Log level";
+    };
+  };
+
+  config = lib.mkIf cfg.enable {
+    systemd.services.nextcloud-notify_push = let
+      nextcloudUrl = "http${lib.optionalString config.services.nextcloud.https "s"}://${config.services.nextcloud.hostName}";
+    in {
+      description = "Push daemon for Nextcloud clients";
+      documentation = [ "https://github.com/nextcloud/notify_push" ];
+      after = [ "phpfpm-nextcloud.service" ];
+      wantedBy = [ "multi-user.target" ];
+      environment = {
+        NEXTCLOUD_URL = nextcloudUrl;
+        SOCKET_PATH = cfg.socketPath;
+        LOG = cfg.logLevel;
+      };
+      postStart = ''
+        ${config.services.nextcloud.occ}/bin/nextcloud-occ notify_push:setup ${nextcloudUrl}/push
+      '';
+      serviceConfig = {
+        ExecStart = "${cfg.package}/bin/notify_push --glob-config ${config.services.nextcloud.datadir}/config/config.php";
```

Comment:
it seems like notify_push got sadly some problem with the way we try to bring passwords to the nextclouds config file, which is used when a [dbpassFile is given](https://github.com/NixOS/nixpkgs/blob/95ea5bb25473fd2bf2f19d8f68444712027ad5eb/nixos/modules/services/web-apps/nextcloud.nix#L834).

I got the following error
```
Dez 16 00:20:39 tank notify_push[42751]: Error: php_literal_parser::unexpected_token
Dez 16 00:20:39 tank notify_push[42751]:   × Error while parsing nextcloud config.php
Dez 16 00:20:39 tank notify_push[42751]:   ╰─▶ Error while parsing '/persist/var/lib/nextcloud/config/
Dez 16 00:20:39 tank notify_push[42751]:       override.config.php':
Dez 16 00:20:39 tank notify_push[42751]:       No valid token found, expected one of boolean literal, integer literal,
Dez 16 00:20:39 tank notify_push[42751]:       float literal, string literal, 'null', 'array' or '['
Dez 16 00:20:39 tank notify_push[42751]:     ╭─[18:1]
Dez 16 00:20:39 tank notify_push[42751]:  18 │
Dez 16 00:20:39 tank notify_push[42751]:  19 │   'dbpassword' => nix_read_secret(
Dez 16 00:20:39 tank notify_push[42751]:     ·                   ┬
Dez 16 00:20:39 tank notify_push[42751]:     ·                   ╰── Expected boolean literal, integer literal, float literal, string literal, 'null', 'array' or '['
Dez 16 00:20:39 tank notify_push[42751]:  20 │   "/run/agenix/nextcloud_db_file"
Dez 16 00:20:39 tank notify_push[42751]:     ╰────
```

whilst the following is the $CONFIG of the override.config.php

```
$CONFIG = [
  'apps_paths' => [
    
    [ 'path' => '/persist/var/lib/nextcloud/apps', 'url' => '/apps', 'writable' => false ],
    [ 'path' => '/persist/var/lib/nextcloud/store-apps', 'url' => '/store-apps', 'writable' => true ],
  ],
  
  'datadirectory' => '/persist/var/lib/nextcloud/data',
  'skeletondirectory' => '',
  
  'log_type' => 'syslog',
  'loglevel' => '2',
  
  'dbname' => 'nextcloud',
  'dbhost' => '/run/postgresql',
  
  'dbuser' => 'nextcloud',
  
  'dbpassword' => nix_read_secret(
  "/run/agenix/nextcloud_db_file"
),

  'dbtype' => 'pgsql',
  'trusted_domains' => ["<some domain>"],
  'trusted_proxies' => ["<some ipv6>"],
  'default_phone_region' => 'DE',
  'profile.enabled' => false,
  
];

```

if not using `--global-config` as argument to enable the loading of the override its passing, but obviously then the overrides are kinda missing.
Not exactly sure what to do here tbh.

Reply:
yeah… I sort of already feared something like this might be an issue. Thanks for reminding me about it.

I'll try to think of a solution. We could maybe take the nextcloud config parameters from `services.nextcloud.config.db*` add them to the environment of the notify push service. That should work.

This whole issue exists, because we use actual php code (the `nix_read_secret` function), while the rust code just uses the `nextcloud_config_parser` rust crate, which doesn't actually evaluate php code, but just parses the file somehow.

Reply:
I thought about it for a bit, but haven't come up with any good idea. Mainly, because I can't even find any documentation for what `DATABASE_URL` accepts as valid urls.

Reply:
They should use sqlx under the hood if I got it right from the code. Then the database format should like 
`<Database type>://<user>:<password>@<host>/<dbname>` eG. `postgres://postgres:password@localhost/test`. Accepted databases should be `postgres`, `mysql`, `mssql`, `sqlite`. 
But honestly I am not completely sure if you tried to ask that.
Beside that we could also add assertion so that this just supports postgres with a passwordless login, which is far from perfect but at least something useable, if the database and the nextcloud are on the same machine.

Reply:
Ah, `sqlx`. That probably helps me, yes. I think, I saw some `sqlite:` urls that didn't contain `//`, so I got confused.

I'll think about it some more and maybe ask some people for ideas.

I'd at least like to support mariadb with passwort auth, because that's what I deploy.

Reply:
https://docs.rs/sqlx/0.5.9/sqlx/sqlite/struct.SqliteConnectOptions.html
https://docs.rs/sqlx/0.5.9/sqlx/postgres/struct.PgConnectOptions.html
https://docs.rs/sqlx/0.5.9/sqlx/mysql/struct.MySqlConnectOptions.html

There should be all things with examples which are accepted my the connection options.
The `sqlite::memory::` and `sqlite:data.db` are (on my opinion) 2 super dump short cuts and are allowed as the underlying implementation accept them. As there is a respective "full uri" alternative, at least on my opinion i would ignore them for this.
Beside the needs to translate the postgres db type, we should (as far as i see) be able to map the nextcloud configs directly to a database url for the environment variable.